flink-end-to-end-tests/flink-streaming-kinesis-test/src/main/java/org/apache/flink/streaming/kinesis/test/KinesisExample.java

```diff
@@ -0,0 +1,157 @@
+ * watermarks directly in the Flink Kinesis consumer. For demonstration purposes, it is assumed that
+ * the String messages formatted as a (word,frequency,timestamp) tuple.
+ *
+ * <p>Example usage:
+ * 	--input-topic test-input --output-topic test-output --bootstrap.servers localhost:9092 --zookeeper.connect localhost:2181 --group.id myconsumer
+ */
+public class KinesisExample {
+	private static final Logger LOG = LoggerFactory.getLogger(KinesisExample.class);
+
+	/**
+	 * Interface to the pubsub system for this test.
+	 */
+	interface PubsubClient {
+		void createTopic(String topic, int partitions, Properties props) throws Exception;
+
+		void sendMessage(String topic, String msg);
+
+		List<String> readAllMessages(String streamName) throws Exception;
+	}
+
+	public static void main(String[] args) throws Exception {
+		LOG.info("System properties: {}", System.getProperties());
+		// parse input arguments
+		final ParameterTool parameterTool = ParameterTool.fromArgs(args);
+		StreamExecutionEnvironment env = KafkaExampleUtil.prepareExecutionEnv(parameterTool);
+
+		String inputStream = parameterTool.getRequired("input-stream");
+		String outputStream = parameterTool.getRequired("output-stream");
+
+		PubsubClient pubsub = new KinesisPubsubClient(parameterTool.getProperties());
+		pubsub.createTopic(inputStream, 2, parameterTool.getProperties());
+		pubsub.createTopic(outputStream, 2, parameterTool.getProperties());
+
+		FlinkKinesisConsumer<KafkaEvent> consumer = new FlinkKinesisConsumer<>(
+			inputStream,
+			new KafkaEventSchema(),
+			parameterTool.getProperties());
+		consumer.setPeriodicWatermarkAssigner(new CustomWatermarkExtractor());
+
+		DataStream<KafkaEvent> input = env
+				.addSource(consumer)
+				.keyBy("word")
+				.map(new RollingAdditionMapper());
+
+		Properties producerProperties = new Properties(parameterTool.getProperties());
+		// needs region event when URL is specified
+		producerProperties.put(ConsumerConfigConstants.AWS_REGION, "us-east-1");
+
+		// KPL does not recognize endpoint URL..
+		String kinesisUrl = producerProperties.getProperty(ConsumerConfigConstants.AWS_ENDPOINT);
+		if (kinesisUrl != null) {
+			URL url = new URL(kinesisUrl);
+			producerProperties.put("KinesisEndpoint", url.getHost());
+			producerProperties.put("KinesisPort", Integer.toString(url.getPort()));
+			producerProperties.put("VerifyCertificate", "false");
+		}
+
+		FlinkKinesisProducer<KafkaEvent> producer = new FlinkKinesisProducer<>(
+			new KafkaEventSchema(),
+			producerProperties);
+		producer.setDefaultStream(outputStream);
+		producer.setDefaultPartition("fakePartition");
+
+		input.addSink(producer);
+
+		final AtomicReference<Exception> caughtException = new AtomicReference<>();
+
+		// TODO: Flink thinks nothing got executed when using different thread
+		// TODO: cancel job before stopping Kinesalite
+		Thread executeThread =
+			new Thread(
+				() -> {
+					try {
+						env.execute();
```

Comment:
@tillrohrmann the test currently fails (after it actually succeeded) because the job isn't cancelled and generates exceptions in the log (after Kinesalite was terminated). I could not quite figure out how I can accommodate both the verification and execution within the main method.

Reply:
This issue is resolved by stopping the flink cluster before Kinesalite in the script. But please see below for need for System.exit(0) to bypass the CLI job check. Is there a better way to fork the job and have main run the test logic?

Reply:
This check for exceptions in the log is a piece of questionable design that is hopefully improved in the future. I think so far, we added certain exceptions that can safely be ignored to a white-list in the script function that checks the logs for exceptions. So you could add the Kinesalite exception there to not have it fail your test. If the exception is of a very generic type ... this is where the questionable design part becomes problematic.

Reply:
This is no longer an issue. The exceptions are avoided with graceful shutdown, terminating the job prior to Kinesalite.